crates/core/component/auction/src/component/action_handler/dutch/schedule.rs

```diff
@@ -1,18 +1,98 @@
-use anyhow::Result;
+use crate::auction::dutch::DutchAuctionDescription;
+use crate::component::AuctionStoreRead;
+use anyhow::{ensure, Result};
 use async_trait::async_trait;
 use cnidarium::StateWrite;
 use cnidarium_component::ActionHandler;
+use penumbra_sct::component::clock::EpochRead; // AuctionRead?
 
-use crate::auction::dutch::actions::ActionDutchAuctionSchedule;
+use crate::auction::dutch::ActionDutchAuctionSchedule;
+use crate::component::DutchAuctionManager;
 
 #[async_trait]
 impl ActionHandler for ActionDutchAuctionSchedule {
     type CheckStatelessContext = ();
     async fn check_stateless(&self, _context: ()) -> Result<()> {
+        let DutchAuctionDescription {
+            input,
+            output_id,
+            max_output,
+            min_output,
+            start_height,
+            end_height,
+            step_count,
+            nonce: _,
+        } = self.description;
+
+        // Fail fast if the step count is zero
+        ensure!(step_count > 0, "step count MUST be positive (got zero)");
+
+        // Check that we disallow identical input/output ids.
+        ensure!(
+            input.asset_id != output_id,
+            "input id MUST be different from output id"
+        );
+
+        // Check that the `max_output` is greater than the `min_output`
+        ensure!(
+            max_output > min_output,
+            "max_output MUST be greater than min_output"
+        );
+
+        // Check that the max output is greater than zero.
+        ensure!(max_output > 0u128.into(), "max output MUST be positive");
+
+        // Check that the start and end height are valid.
+        ensure!(
+            start_height < end_height,
+            "the start height MUST be strictly less than the end height (got: start={} >= end={})",
+            start_height,
+            end_height
+        );
+
+        // Check that the step count is positive.
+        ensure!(
+            step_count > 0,
+            "step count MUST be greater than zero (got: {step_count})"
+        );
+
+        // Check that the step count is less than 1000.
+        ensure!(
+            step_count <= 1000,
+            "the dutch auction step count MUST be less than 1000 (got: {step_count})",
+        );
+
+        // Check that height delta is a multiple of `step_count`.
+        let block_window = end_height - start_height;
```

Comment:
if we wrote this with `checked_sub` one of the invariant checks above would be addressed naturally here. i would "like" that but it is fine as written :)

Reply:
i think having both is better, good idea